x/ccv/provider/keeper/double_vote.go

```diff
@@ -0,0 +1,115 @@
+package keeper
+
+import (
+	"bytes"
+	"fmt"
+
+	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/interchain-security/v2/x/ccv/provider/types"
+	ccvtypes "github.com/cosmos/interchain-security/v2/x/ccv/types"
+	tmprotocrypto "github.com/tendermint/tendermint/proto/tendermint/crypto"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+// HandleConsumerDoubleVoting verifies a double voting evidence for a given a consumer chain and,
+// if successful, executes the the jailing and the tombstoning of the malicious validator.
+func (k Keeper) HandleConsumerDoubleVoting(ctx sdk.Context, evidence *tmtypes.DuplicateVoteEvidence, chainID string) error {
+	// get the validator's consensus address on the provider
+	providerAddr := k.GetProviderAddrFromConsumerAddr(
+		ctx,
+		chainID,
+		types.NewConsumerConsAddress(sdk.ConsAddress(evidence.VoteA.ValidatorAddress.Bytes())),
+	)
+
+	// get the consumer chain public key assigned to the validator
+	consuPubkey, ok := k.GetValidatorConsumerPubKey(ctx, chainID, providerAddr)
+	if !ok {
+		return fmt.Errorf("cannot find public key for validator %s and consumer chain %s", providerAddr.String(), chainID)
+	}
+
+	// verifies the double voting evidence using the consumer chain public key
+	if err := k.VerifyDoubleVotingEvidence(ctx, *evidence, chainID, consuPubkey); err != nil {
+		return err
+	}
+
+	// execute the jailing and tombstoning
+	k.JailAndTombstoneValidator(ctx, providerAddr)
+
+	k.Logger(ctx).Info(
+		"confirmed equivocation",
+		"byzantine validator address", providerAddr,
+	)
+
+	return nil
+}
+
+// VerifyDoubleVotingEvidence verifies a double voting evidence
+// for a given chain id and validator public key
+func (k Keeper) VerifyDoubleVotingEvidence(
+	ctx sdk.Context,
+	evidence tmtypes.DuplicateVoteEvidence,
+	chainID string,
+	pubkey tmprotocrypto.PublicKey,
+) error {
+	// check evidence age duration
+	maxAgeDuration := ctx.ConsensusParams().Evidence.MaxAgeDuration
+	ageDuration := ctx.BlockHeader().Time.Sub(evidence.Time())
```

Comment:
`evidence.Time()` uses the `Timestamp` from `DuplicateVoteEvidence` that in my understanding could be easily faked? If that's the case, someone could send old evidence with a recent timestamp.
Would it make sense to get the time out of one of the votes instead?

Reply:
Good catch! Changed to use the vote timestamp instead.